Gameplay/entities/player/CombatManager.cpp

```diff
@@ -0,0 +1,576 @@
+	ComponentTransform* bullet_transform = bullet->GetTransform();
+	const float bullet_offset = 1.25f;
+	float3 emitter_direction = stats.emitter_transform->GetFront().Normalized();
+	float3 emitter_position = stats.emitter_transform->GetGlobalPosition() + emitter_direction * bullet_offset;
+	
+	stats.prev_position = emitter_position;
+
+	stats.direction = emitter_direction;
+	bullet_transform->SetGlobalPosition(emitter_position);
+	// If the bullet is out of culling nothing its updating its transform unless selected in engine
+	bullet_transform->GetLocalPosition();
+}
+
+Hachiko::Scripting::EnemyController* Hachiko::Scripting::CombatManager::FindBulletClosestEnemyHit(GameObject* bullet, float bullet_size, LineSegment& trajectory, float& closest_hit)
+{
+	EnemyController* hit_target = nullptr;
+	
+	GameObject* agent_container = game_object->scene_owner->GetRoot()->GetFirstChildWithName("Enemies");
+	if (!agent_container)
+	{
+		return nullptr;
+	}
+
+	float3 bullet_position = bullet->GetTransform()->GetGlobalPosition();
+
+
+	std::vector<GameObject*> enemies = agent_container->children;
+	for (int i = 0; i < enemies.size(); ++i)
+	{
+		GameObject* enemy = enemies[i];
+		if (!enemy->IsActive())
+		{
+			continue;
+		}
+		
+		ComponentAgent* agent = enemy->GetComponent<ComponentAgent>();
+		if (!agent)
+		{
+			continue;
+		}
+
+		float3 enemy_position = enemy->GetTransform()->GetGlobalPosition();
+		float enemy_radius = agent->GetRadius();
+		Sphere hitbox = Sphere(enemy_position, enemy_radius + bullet_size);
+
+		if (hitbox.Intersects(hitbox))
+		{
+			EnemyController* enemy_controller = enemy->GetComponent<EnemyController>();
+			float hit_distance = bullet_position.Distance(enemy_position);
+			if (enemy_controller && hit_distance < closest_hit)
+			{
+				
+				closest_hit = hit_distance;
+				hit_target = enemy_controller;
+			}
+		}
+	}
+	return hit_target;
+}
+
+Hachiko::Scripting::CrystalExplosion* Hachiko::Scripting::CombatManager::FindBulletClosestObstacleHit(GameObject* bullet, float bullet_size, LineSegment& trajectory, float& closest_hit)
+{
+	CrystalExplosion* hit_target = nullptr;
+
+	GameObject* obstacle_container = game_object->scene_owner->GetRoot()->GetFirstChildWithName("Crystals");
+	if (!obstacle_container)
+	{
+		return nullptr;
+	}
+
+	float3 bullet_position = bullet->GetTransform()->GetGlobalPosition();
+
+
+	std::vector<GameObject*> obstacles = obstacle_container->children;
+	for (int i = 0; i < obstacles.size(); ++i)
+	{
+		GameObject* obstacle = obstacles[i];
+		if (!obstacle->IsActive())
+		{
+			continue;
+		}
+
+		ComponentObstacle* obstacle_cpomponent = obstacle->GetComponent<ComponentObstacle>();
+		if (!obstacle_cpomponent)
+		{
+			continue;
+		}
+		
+		float3 obstacle_position = obstacle->GetTransform()->GetGlobalPosition();
+		// Cylinder obstacles ignore z
+		float obstacle_radius = obstacle_cpomponent->GetSize().x;
+		Sphere hitbox = Sphere(obstacle_position, obstacle_radius + bullet_size);
```

Comment:
I like using the hitbox instead of the distance but if at the end we compute the distance shouldn't we use that?